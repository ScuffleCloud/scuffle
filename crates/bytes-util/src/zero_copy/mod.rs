//! Zero-copy types.

mod reader;
pub use reader::*;

mod serde;
pub use serde::*;
