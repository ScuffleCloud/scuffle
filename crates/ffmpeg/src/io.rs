mod input;
mod internal;
mod output;

/// A module that contains the channel implementation for io operations.
#[cfg(feature = "channel")]
pub mod channel;

pub use input::*;
pub use output::*;
