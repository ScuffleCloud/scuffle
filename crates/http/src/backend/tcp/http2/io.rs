#![allow(dead_code)]

// TODO: Use BDP logic for bandwidth limiting & keep alive logic

use std::pin::Pin;
use std::sync::Arc;
use std::task::{Context, Poll};

use bytes::Bytes;
use spin::Mutex;

use super::ResultErrorExt;
use crate::error::{ErrorConfig, ErrorScope, ErrorSeverity};
use crate::Error;

#[derive(Clone)]
/// Shared state is shared between all streams on the same connection.
pub(crate) struct Http2SharedState {
	inner: Arc<Mutex<Http2SharedStateInner>>,
}

impl Http2SharedState {
	pub(crate) fn new(ping_pong: h2::PingPong) -> Self {
		Self {
			inner: Arc::new(Mutex::new(Http2SharedStateInner { ping_pong })),
		}
	}
}

struct Http2SharedStateInner {
	ping_pong: h2::PingPong,
}

#[derive(Clone, Copy)]
enum State {
	SizeKnown(u64),
	SizeUnknown,
	BodyComplete,
	TrailerComplete,
}

pin_project_lite::pin_project! {
	pub struct RecvStream {
		#[pin]
		io: h2::RecvStream,
		#[pin]
		state: State,
		shared: Http2SharedState,
	}
}

impl RecvStream {
	pub(crate) fn new(io: h2::RecvStream, size: Option<u64>, shared: Http2SharedState) -> Self {
		Self {
			io,
			state: size.map(State::SizeKnown).unwrap_or(State::SizeUnknown),
			shared,
		}
	}
}

impl http_body::Body for RecvStream {
	type Data = Bytes;
	type Error = crate::Error;

	fn poll_frame(
		self: Pin<&mut Self>,
		cx: &mut Context<'_>,
	) -> Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>> {
		let mut this = self.project();

		if matches!(*this.state, State::TrailerComplete) {
			return Poll::Ready(None);
		}

		if !matches!(*this.state, State::BodyComplete) {
			match this.io.poll_data(cx) {
				Poll::Pending => return Poll::Pending,
				Poll::Ready(Some(data)) => {
					let data = data.with_config(ErrorConfig {
						severity: ErrorSeverity::Unknown,
						scope: ErrorScope::Request,
						context: "http2 body data",
					})?;

					return Poll::Ready(Some(Ok(http_body::Frame::data(data))));
				}
				Poll::Ready(None) => {
					*this.state = State::BodyComplete;
				}
			}
		}

		match this.io.poll_trailers(cx) {
			Poll::Pending => Poll::Pending,
			Poll::Ready(Ok(None)) => {
				*this.state = State::TrailerComplete;
				Poll::Ready(None)
			}
			Poll::Ready(Ok(Some(trailers))) => {
				*this.state = State::TrailerComplete;
				Poll::Ready(Some(Ok(http_body::Frame::trailers(trailers))))
			}
			Poll::Ready(Err(e)) => {
				*this.state = State::BodyComplete;
				Poll::Ready(Some(Err(Error::from(e).with_config(ErrorConfig {
					severity: ErrorSeverity::Unknown,
					scope: ErrorScope::Request,
					context: "http2 body trailers",
				}))))
			}
		}
	}

	fn is_end_stream(&self) -> bool {
		self.io.is_end_stream() || matches!(self.state, State::TrailerComplete)
	}

	fn size_hint(&self) -> http_body::SizeHint {
		match self.state {
			State::SizeKnown(size) => http_body::SizeHint::with_exact(size),
			State::SizeUnknown => http_body::SizeHint::default(),
			State::BodyComplete | State::TrailerComplete => http_body::SizeHint::with_exact(0),
		}
	}
}
