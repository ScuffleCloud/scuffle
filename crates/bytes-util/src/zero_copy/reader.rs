//! Zero-copy reader types.

use std::io::{self, Read};

use crate::BytesCow;

/// A trait for zero-copy readers.
pub trait ZeroCopyReader<'a> {
    /// Attempts to read a specified number of bytes from the reader without copying.
    ///
    /// This function does not guarantee that no copying will occur.
    /// Some implementations can't avoid copying.
    fn try_read(&mut self, size: usize) -> Result<BytesCow<'a>, io::Error>;

    /// Returns a standard [`io::Read`] interface for the reader.
    fn as_std(&mut self) -> impl io::Read;

    /// Limits the number of bytes that can be read from the reader.
    fn take(self, limit: usize) -> Take<Self>
    where
        Self: Sized,
    {
        Take::new(self, limit)
    }

    /// Reads all remaining bytes from the reader.
    fn try_read_to_end(&mut self) -> Result<BytesCow<'a>, io::Error> {
        let mut buf = Vec::new();
        self.as_std().read_to_end(&mut buf)?;
        Ok(BytesCow::from_vec(buf))
    }
}

impl<'a, T: ZeroCopyReader<'a>> ZeroCopyReader<'a> for &mut T {
    fn try_read(&mut self, size: usize) -> Result<BytesCow<'a>, io::Error> {
        (*self).try_read(size)
    }

    fn as_std(&mut self) -> impl io::Read {
        (*self).as_std()
    }
}

/// A zero-copy reader that wraps a [`bytes::Buf`].
pub struct BytesBuf<B>(B);

impl<B: bytes::Buf> From<B> for BytesBuf<B> {
    fn from(buf: B) -> Self {
        Self(buf)
    }
}

impl<'a, B: bytes::Buf> ZeroCopyReader<'a> for BytesBuf<B> {
    fn try_read(&mut self, size: usize) -> Result<BytesCow<'a>, io::Error> {
        if self.0.remaining() < size {
            return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "Not enough data"));
        }

        Ok(BytesCow::from_bytes(self.0.copy_to_bytes(size)))
    }

    fn as_std(&mut self) -> impl io::Read {
        bytes::Buf::reader(&mut self.0)
    }
}

/// A zero-copy reader that wraps a [`std::io::Read`].
///
/// This implementation is not zero-copy and will always copy the data into a new buffer.
/// It is not possible to implement zero-copy reading for [`std::io::Read`]
/// because it does not provide a way to access the underlying buffer directly.
pub struct IoRead<R>(R);

impl<R: io::Read> From<R> for IoRead<R> {
    fn from(reader: R) -> Self {
        Self(reader)
    }
}

impl<'a, R: io::Read> ZeroCopyReader<'a> for IoRead<R> {
    fn try_read(&mut self, size: usize) -> Result<BytesCow<'a>, io::Error> {
        let mut buf = vec![0; size];
        self.0.read_exact(&mut buf)?;
        Ok(BytesCow::from_vec(buf))
    }

    fn as_std(&mut self) -> impl io::Read {
        &mut self.0
    }
}

/// A zero-copy reader that wraps a byte slice (`&[u8]`).
pub struct Slice<'a>(io::Cursor<&'a [u8]>);

impl<'a> From<&'a [u8]> for Slice<'a> {
    fn from(slice: &'a [u8]) -> Self {
        Self(io::Cursor::new(slice))
    }
}

impl<'a> ZeroCopyReader<'a> for Slice<'a> {
    fn try_read(&mut self, size: usize) -> Result<BytesCow<'a>, io::Error> {
        let start = self.0.position() as usize;
        let end = start + size;

        if end > self.0.get_ref().len() {
            return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "Not enough data"));
        }

        let slice = &self.0.get_ref()[start..end];
        self.0.set_position(end as u64);
        Ok(BytesCow::from_slice(slice))
    }

    fn as_std(&mut self) -> impl io::Read {
        &mut self.0
    }
}

/// Zero-copy reader that limits the number of bytes read.
///
/// This is equivalent to [`std::io::Take`], but it implements [`ZeroCopyReader`].
pub struct Take<R> {
    inner: R,
    limit: usize,
}

struct TakeAsStd<'a, R> {
    inner: R,
    limit: &'a mut usize,
}

impl<R: io::Read> io::Read for TakeAsStd<'_, R> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        if *self.limit == 0 {
            return Ok(0);
        }

        let size = std::cmp::min(buf.len(), *self.limit);
        let n = self.inner.read(&mut buf[..size])?;
        *self.limit -= n;
        Ok(n)
    }
}

impl<R> Take<R> {
    /// Creates a new [`Take`] reader that limits the number of bytes read.
    pub fn new(inner: R, limit: usize) -> Self {
        Self { inner, limit }
    }

    /// Returns the underlying reader.
    pub fn into_inner(self) -> R {
        self.inner
    }
}

impl<'a, R> ZeroCopyReader<'a> for Take<R>
where
    R: ZeroCopyReader<'a>,
{
    fn as_std(&mut self) -> impl io::Read {
        TakeAsStd {
            inner: self.inner.as_std(),
            limit: &mut self.limit,
        }
    }

    fn try_read(&mut self, size: usize) -> Result<BytesCow<'a>, io::Error> {
        let size = std::cmp::min(size, self.limit);
        let result = self.inner.try_read(size)?;
        self.limit -= result.len();
        Ok(result)
    }
}

#[cfg(test)]
#[cfg_attr(all(test, coverage_nightly), coverage(off))]
mod tests {
    use super::{Slice, ZeroCopyReader};

    #[test]
    fn take_and_read_to_end() {
        let data = b"Hello, world!";
        let mut reader = Slice::from(&data[..]).take(5);
        assert_eq!(reader.try_read_to_end().unwrap(), b"Hello");
    }
}
