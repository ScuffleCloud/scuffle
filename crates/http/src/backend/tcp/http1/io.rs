use std::pin::Pin;
use std::task::{Context, Poll};

use bytes::{Buf, BufMut, Bytes, BytesMut};
use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt, BufReader, BufWriter};
use tokio::sync::mpsc;

use super::{proto, ResultErrorExt};
use crate::error::{ErrorConfig, ErrorScope, ErrorSeverity};

#[derive(Clone, Copy)]
enum State {
	SizeKnown(u64),
	SizeUnknown,
	BodyComplete,
	TrailerComplete,
}

pin_project_lite::pin_project! {
	pub struct RecvStream {
		#[pin]
		state: State,
		#[pin]
		inner: mpsc::Receiver<http_body::Frame<Bytes>>,
	}
}

impl RecvStream {
	pub fn new(inner: mpsc::Receiver<http_body::Frame<Bytes>>, size: Option<u64>) -> Self {
		Self {
			state: size.map_or(State::SizeUnknown, State::SizeKnown),
			inner,
		}
	}
}

impl http_body::Body for RecvStream {
	type Data = Bytes;
	type Error = std::io::Error;

	fn poll_frame(
		self: Pin<&mut Self>,
		cx: &mut Context<'_>,
	) -> Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>> {
		let mut this = self.project();

		let remaining = match *this.state {
			State::SizeUnknown => None,
			State::SizeKnown(remaining) => Some(remaining),
			State::BodyComplete => None,
			State::TrailerComplete => return Poll::Ready(None),
		};

		match this.inner.poll_recv(cx) {
			Poll::Pending => Poll::Pending,
			Poll::Ready(Some(frame)) => {
				match frame.data_ref() {
					Some(data) => {
						if let Some(remaining) = remaining {
							if remaining >= data.len() as u64 {
								*this.state = State::SizeKnown(remaining - data.len() as u64);
							} else {
								*this.state = State::BodyComplete;
							}
						}
					}
					None => {
						// If we get trailers, the body is done.
						*this.state = State::TrailerComplete;
						return Poll::Ready(Some(Ok(frame)));
					}
				}

				Poll::Ready(Some(Ok(frame)))
			}
			Poll::Ready(None) => {
				if matches!(*this.state, State::BodyComplete | State::TrailerComplete) {
					Poll::Ready(None)
				} else {
					Poll::Ready(Some(Err(std::io::Error::new(
						std::io::ErrorKind::UnexpectedEof,
						"Request too short",
					))))
				}
			}
		}
	}

	fn is_end_stream(&self) -> bool {
		matches!(self.state, State::BodyComplete | State::TrailerComplete)
	}

	fn size_hint(&self) -> http_body::SizeHint {
		match self.state {
			State::SizeKnown(remaining) => http_body::SizeHint::with_exact(remaining),
			State::SizeUnknown => http_body::SizeHint::new(),
			State::BodyComplete => http_body::SizeHint::with_exact(0),
			State::TrailerComplete => http_body::SizeHint::with_exact(0),
		}
	}
}

pub enum BodyEncoding {
	Chunked { max_headers: usize },
	ContentLength(u64),
}

pub enum BodyParser<'a, R: AsyncRead + Unpin> {
	Empty,
	Reader {
		reader: &'a mut BufReader<R>,
		body_sender: mpsc::Sender<http_body::Frame<Bytes>>,
		encoding: BodyEncoding,
	},
}

impl<'a, R: AsyncRead + Unpin> BodyParser<'a, R> {
	pub fn empty() -> Self {
		Self::Empty
	}

	pub fn reader(
		reader: &'a mut BufReader<R>,
		body_sender: mpsc::Sender<http_body::Frame<Bytes>>,
		encoding: BodyEncoding,
	) -> Self {
		Self::Reader {
			reader,
			body_sender,
			encoding,
		}
	}

	pub async fn read(self) -> Result<(), crate::Error> {
		match self {
			Self::Empty => Ok(()),
			Self::Reader {
				reader,
				body_sender,
				encoding,
			} => match encoding {
				BodyEncoding::Chunked { max_headers } => {
					// Loop for each chunk.
					loop {
						let chunk_size = httparse::parse_chunk_size(reader.buffer()).map_err(|_| ErrorConfig {
							severity: ErrorSeverity::Debug,
							scope: ErrorScope::Connection,
							context: "parse chunk size",
						})?;

						let mut chunk_size = match chunk_size {
							httparse::Status::Complete((idx, size)) => {
								reader.consume(idx);
								size
							}
							httparse::Status::Partial => {
								if reader.buffer().len() > 128 {
									return Err(ErrorConfig {
										severity: ErrorSeverity::Debug,
										scope: ErrorScope::Connection,
										context: "chunk header too large",
									}
									.into());
								}

								reader.fill_buf().await?;
								continue;
							}
						};

						// A zero-length chunk indicates the end of the body.
						if chunk_size == 0 {
							break;
						}

						while (reader.buffer().len() as u64) < chunk_size {
							if chunk_size != 0 && reader.buffer().is_empty() {
								reader.fill_buf().await?;
								if reader.buffer().is_empty() {
									return Err(ErrorConfig {
										severity: ErrorSeverity::Debug,
										scope: ErrorScope::Connection,
										context: "Request too short",
									}
									.into());
								}
							}

							let chunk = Bytes::copy_from_slice(reader.buffer());
							reader.consume(chunk.len());
							chunk_size -= chunk.len() as u64;
							body_sender.send(http_body::Frame::data(chunk)).await.ok();
						}

						if chunk_size != 0 {
							let chunk = Bytes::copy_from_slice(&reader.buffer()[..chunk_size as usize]);
							reader.consume(chunk.len());
							body_sender.send(http_body::Frame::data(chunk)).await.ok();
						}
					}

					loop {
						let Some((idx, headers)) = proto::parse_headers(reader.buffer(), max_headers)? else {
							if reader.buffer().len() > 1024 {
								return Err(ErrorConfig {
									severity: ErrorSeverity::Debug,
									scope: ErrorScope::Connection,
									context: "trailers too large",
								}
								.into());
							}

							let size = reader.buffer().len();
							reader.fill_buf().await?;
							if reader.buffer().len() == size {
								return Err(ErrorConfig {
									severity: ErrorSeverity::Debug,
									scope: ErrorScope::Connection,
									context: "Request too short",
								}
								.into());
							}

							continue;
						};

						reader.consume(idx);
						body_sender.send(http_body::Frame::trailers(headers)).await.ok();
						break;
					}

					Ok(())
				}
				BodyEncoding::ContentLength(mut remaining) => {
					while remaining > 0 {
						reader.fill_buf().await?;

						if reader.buffer().is_empty() {
							return Err(ErrorConfig {
								severity: ErrorSeverity::Debug,
								scope: ErrorScope::Connection,
								context: "Request too short",
							}
							.into());
						}

						let should_consume = reader.buffer().len().min(remaining as usize);
						let chunk = Bytes::copy_from_slice(&reader.buffer()[..should_consume]);

						remaining -= chunk.len() as u64;
						reader.consume(chunk.len());

						// I am not sure if we should ignore errors here.
						body_sender.send(http_body::Frame::data(chunk)).await.ok();
					}

					Ok(())
				}
			},
		}
	}
}

pub async fn copy_body<E: Into<crate::Error>>(
	body: impl http_body::Body<Error = E>,
	writer: &mut BufWriter<impl AsyncWrite + Unpin>,
) -> Result<(), crate::Error> {
	let mut size = body.size_hint().exact();

	let mut body = std::pin::pin!(body);

	let mut chunked_buf = BytesMut::new();

	while let Some(frame) = std::future::poll_fn(|cx| body.as_mut().poll_frame(cx)).await {
		chunked_buf.clear();

		match frame {
			Ok(mut frame) => {
				if let Some(data) = frame.data_mut() {
					match &mut size {
						Some(remaining) => {
							while *remaining > 0 && data.has_remaining() {
								let chunk = data.chunk();
								let size = chunk.len().min(*remaining as usize);
								writer.write_all(&chunk[..size]).await.with_context("body write")?;
								*remaining -= size as u64;
								data.advance(size);
							}
						}
						None => {
							let size = data.remaining();

							// TODO we should not use std::fmt here as this is very slow & does small heap
							// allocations.
							let hex_size = format!("{:x}", size as u64);

							chunked_buf.put_slice(hex_size.as_bytes());
							chunked_buf.put_slice(b"\r\n");

							writer
								.write_all(chunked_buf.as_ref())
								.await
								.with_context("body write chunk header")?;

							// We should take advantage of vectored writes here maybe?
							while data.has_remaining() {
								let chunk = data.chunk();
								writer.write_all(chunk).await.with_context("body write chunk")?;
								data.advance(chunk.len());
							}

							writer.write_all(b"\r\n").await.with_context("body write chunk trailer")?;
							writer.flush().await.with_context("body flush")?;
						}
					}
				} else if let Some(trailers) = frame.trailers_ref() {
					if size.is_some() {
						return Err(ErrorConfig {
							severity: ErrorSeverity::Debug,
							scope: ErrorScope::Connection,
							context: "Trailers specified with non-chunked encoding",
						}
						.into());
					}

					chunked_buf.put_slice(b"0\r\n");
					proto::write_headers(&mut chunked_buf, trailers);
					writer
						.write_all(chunked_buf.as_ref())
						.await
						.with_context("body write chunk trailers")?;
					return Ok(());
				}
			}
			Err(e) => {
				return Err(e.into().with_config(ErrorConfig {
					severity: ErrorSeverity::Error,
					scope: ErrorScope::Connection,
					context: "copy body",
				}));
			}
		}

		if size.is_some_and(|s| s == 0) {
			break;
		}
	}

	if size.is_none() {
		writer
			.write_all(b"0\r\n\r\n")
			.await
			.with_context("body write chunk trailers")?;
	}

	Ok(())
}
