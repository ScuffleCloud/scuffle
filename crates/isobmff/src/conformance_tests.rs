#![cfg(test)]
#![cfg_attr(all(test, coverage_nightly), coverage(off))]

use std::collections::BTreeMap;
use std::io::BufReader;
use std::path::PathBuf;

use pretty_assertions::assert_eq;
use scuffle_bytes_util::zero_copy::{Deserialize, Serialize};

use crate::{IsoSized, IsobmffFile};

#[derive(Debug, serde_derive::Deserialize)]
struct MetadataFile {
    file_metadata: BTreeMap<String, FileMetadata>,
}

#[derive(Debug, serde_derive::Deserialize)]
struct FileMetadata {
    published: bool,
}

#[test]
fn conformance_files() {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../assets")
        .join("isobmff_conformance");

    let metadata_file: MetadataFile =
        serde_json::from_reader(std::fs::File::open(dir.join("files.json")).expect("failed to open metadata file"))
            .expect("failed to deserialize metadata file");

    for (file_name, _) in metadata_file
        .file_metadata
        .into_iter()
        .filter(|(n, m)| !n.ends_with(".zip") && m.published)
    {
        println!("testing {file_name}");
        let mut file = std::fs::File::open(dir.join("files").join(&file_name)).expect("failed to open file");
        let reader = scuffle_bytes_util::zero_copy::IoRead::from(BufReader::new(&mut file));

        let isobmff_file = IsobmffFile::deserialize(reader).expect("failed to deserialize file");

        let file_size = file.metadata().expect("failed to read metadata").len() as usize;
        assert!(
            isobmff_file.size() <= file_size,
            "file size mismatch: {} (serialized size) > {} (file size)",
            isobmff_file.size(),
            file_size
        );

        let mut serialized = Vec::new();
        isobmff_file.serialize(&mut serialized).expect("failed to serialize file");
        std::fs::write(
            dir.join("output").join(file_name.split('/').last().unwrap_or(&file_name)),
            &serialized,
        )
        .expect("failed to write serialized file");
        let redeserialized_file = IsobmffFile::deserialize(scuffle_bytes_util::zero_copy::Slice::from(&serialized[..]))
            .expect("failed to deserialize serialized file");
        assert_eq!(isobmff_file, redeserialized_file, "file content mismatch");
    }
}
