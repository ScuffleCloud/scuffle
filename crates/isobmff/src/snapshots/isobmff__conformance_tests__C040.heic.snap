---
source: crates/isobmff/src/conformance_tests.rs
expression: redeserialized_file
snapshot_kind: text
---
IsobmffFile {
    ftyp: Some(
        FileTypeBox {
            major_brand: Brand([109, 105, 102, 49]),
            minor_version: 0,
            compatible_brands: [
                Brand([109, 105, 102, 49]),
                Brand([104, 101, 105, 99]),
            ],
        },
    ),
    etyp: [],
    otyp: [],
    pdin: None,
    moov: None,
    moof: [],
    mdat: [
        MediaDataBox {
            data.len: 155793,
        },
    ],
    imda: [],
    meta: Some(
        MetaBox {
            full_header: FullBoxHeader {
                version: 0,
                flags: U24Be(
                    0,
                ),
            },
            hdlr: HandlerBox {
                full_header: FullBoxHeader {
                    version: 0,
                    flags: U24Be(
                        0,
                    ),
                },
                pre_defined: 0,
                handler_type: HandlerType([112, 105, 99, 116]),
                reserved: [
                    0,
                    0,
                    0,
                ],
                name: Utf8String(
                    "",
                ),
            },
            dinf: None,
            iloc: Some(
                ItemLocationBox {
                    full_header: FullBoxHeader {
                        version: 0,
                        flags: U24Be(
                            0,
                        ),
                    },
                    offset_size: 4,
                    length_size: 4,
                    base_offset_size: 4,
                    index_size: 0,
                    item_count: Some(
                        5,
                    ),
                    items: [
                        ItemLocationBoxItem {
                            item_id: Some(
                                1002,
                            ),
                            construction_method: None,
                            data_reference_index: 0,
                            base_offset: 1027,
                            extent_count: 1,
                            extents: [
                                ItemLocationBoxExtent {
                                    item_reference_index: None,
                                    extent_offset: 16,
                                    extent_length: 19260,
                                },
                            ],
                        },
                        ItemLocationBoxItem {
                            item_id: Some(
                                1005,
                            ),
                            construction_method: None,
                            data_reference_index: 0,
                            base_offset: 1027,
                            extent_count: 1,
                            extents: [
                                ItemLocationBoxExtent {
                                    item_reference_index: None,
                                    extent_offset: 19276,
                                    extent_length: 19379,
                                },
                            ],
                        },
                        ItemLocationBoxItem {
                            item_id: Some(
                                1008,
                            ),
                            construction_method: None,
                            data_reference_index: 0,
                            base_offset: 1027,
                            extent_count: 1,
                            extents: [
                                ItemLocationBoxExtent {
                                    item_reference_index: None,
                                    extent_offset: 38655,
                                    extent_length: 19545,
                                },
                            ],
                        },
                        ItemLocationBoxItem {
                            item_id: Some(
                                1011,
                            ),
                            construction_method: None,
                            data_reference_index: 0,
                            base_offset: 1027,
                            extent_count: 1,
                            extents: [
                                ItemLocationBoxExtent {
                                    item_reference_index: None,
                                    extent_offset: 58200,
                                    extent_length: 19868,
                                },
                            ],
                        },
                        ItemLocationBoxItem {
                            item_id: Some(
                                1014,
                            ),
                            construction_method: None,
                            data_reference_index: 0,
                            base_offset: 1027,
                            extent_count: 1,
                            extents: [
                                ItemLocationBoxExtent {
                                    item_reference_index: None,
                                    extent_offset: 78068,
                                    extent_length: 77741,
                                },
                            ],
                        },
                    ],
                },
            ),
            ipro: None,
            iinf: Some(
                ItemInfoBox {
                    full_header: FullBoxHeader {
                        version: 0,
                        flags: U24Be(
                            0,
                        ),
                    },
                    entry_count: 5,
                    item_infos: [
                        ItemInfoEntry {
                            full_header: FullBoxHeader {
                                version: 2,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            item_id: Some(
                                1002,
                            ),
                            item_protection_index: 0,
                            item_type: [
                                104,
                                118,
                                99,
                                49,
                            ],
                            item_name: Utf8String(
                                "HEVC Image",
                            ),
                            item: None,
                            extension_type: None,
                            extension: None,
                        },
                        ItemInfoEntry {
                            full_header: FullBoxHeader {
                                version: 2,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            item_id: Some(
                                1005,
                            ),
                            item_protection_index: 0,
                            item_type: [
                                104,
                                118,
                                99,
                                49,
                            ],
                            item_name: Utf8String(
                                "HEVC Image",
                            ),
                            item: None,
                            extension_type: None,
                            extension: None,
                        },
                        ItemInfoEntry {
                            full_header: FullBoxHeader {
                                version: 2,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            item_id: Some(
                                1008,
                            ),
                            item_protection_index: 0,
                            item_type: [
                                104,
                                118,
                                99,
                                49,
                            ],
                            item_name: Utf8String(
                                "HEVC Image",
                            ),
                            item: None,
                            extension_type: None,
                            extension: None,
                        },
                        ItemInfoEntry {
                            full_header: FullBoxHeader {
                                version: 2,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            item_id: Some(
                                1011,
                            ),
                            item_protection_index: 0,
                            item_type: [
                                104,
                                118,
                                99,
                                49,
                            ],
                            item_name: Utf8String(
                                "HEVC Image",
                            ),
                            item: None,
                            extension_type: None,
                            extension: None,
                        },
                        ItemInfoEntry {
                            full_header: FullBoxHeader {
                                version: 2,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            item_id: Some(
                                1014,
                            ),
                            item_protection_index: 0,
                            item_type: [
                                104,
                                118,
                                99,
                                49,
                            ],
                            item_name: Utf8String(
                                "HEVC Image",
                            ),
                            item: None,
                            extension_type: None,
                            extension: None,
                        },
                    ],
                },
            ),
            xml: None,
            bxml: None,
            pitm: Some(
                PrimaryItemBox {
                    full_header: FullBoxHeader {
                        version: 0,
                        flags: U24Be(
                            0,
                        ),
                    },
                    item_id: 1014,
                },
            ),
            fiin: None,
            idat: None,
            iref: Some(
                ItemReferenceBox {
                    full_header: FullBoxHeader {
                        version: 0,
                        flags: U24Be(
                            0,
                        ),
                    },
                    references: [
                        SingleItemTypeReferenceBox {
                            large: false,
                            header: BoxHeader {
                                size: Short(
                                    20,
                                ),
                                box_type: FourCc(
                                    "base",
                                ),
                            },
                            from_item_id: 1014,
                            reference_count: 4,
                            to_item_id: [
                                1002,
                                1005,
                                1008,
                                1011,
                            ],
                        },
                    ],
                },
            ),
            iprp: Some(
                ItemPropertiesBox {
                    property_container: ItemPropertyContainerBox {
                        etyp: None,
                        brnd: [],
                        scrb: [],
                        boxes: [
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        108,
                                    ),
                                    box_type: FourCc(
                                        "hvcC",
                                    ),
                                },
                                data.len: 100,
                            },
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        20,
                                    ),
                                    box_type: FourCc(
                                        "ispe",
                                    ),
                                },
                                data.len: 12,
                            },
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        108,
                                    ),
                                    box_type: FourCc(
                                        "hvcC",
                                    ),
                                },
                                data.len: 100,
                            },
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        108,
                                    ),
                                    box_type: FourCc(
                                        "hvcC",
                                    ),
                                },
                                data.len: 100,
                            },
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        108,
                                    ),
                                    box_type: FourCc(
                                        "hvcC",
                                    ),
                                },
                                data.len: 100,
                            },
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        108,
                                    ),
                                    box_type: FourCc(
                                        "hvcC",
                                    ),
                                },
                                data.len: 100,
                            },
                            UnknownBox {
                                header: BoxHeader {
                                    size: Short(
                                        20,
                                    ),
                                    box_type: FourCc(
                                        "ispe",
                                    ),
                                },
                                data.len: 12,
                            },
                        ],
                    },
                    association: [
                        ItemPropertyAssociationBox {
                            full_header: FullBoxHeader {
                                version: 0,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            entry_count: 5,
                            entries: [
                                ItemPropertyAssociationBoxEntry {
                                    item_id: 1002,
                                    association_count: 2,
                                    associations: [
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: true,
                                            property_index: 1,
                                        },
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: false,
                                            property_index: 2,
                                        },
                                    ],
                                },
                                ItemPropertyAssociationBoxEntry {
                                    item_id: 1005,
                                    association_count: 2,
                                    associations: [
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: true,
                                            property_index: 3,
                                        },
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: false,
                                            property_index: 2,
                                        },
                                    ],
                                },
                                ItemPropertyAssociationBoxEntry {
                                    item_id: 1008,
                                    association_count: 2,
                                    associations: [
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: true,
                                            property_index: 4,
                                        },
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: false,
                                            property_index: 2,
                                        },
                                    ],
                                },
                                ItemPropertyAssociationBoxEntry {
                                    item_id: 1011,
                                    association_count: 2,
                                    associations: [
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: true,
                                            property_index: 5,
                                        },
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: false,
                                            property_index: 2,
                                        },
                                    ],
                                },
                                ItemPropertyAssociationBoxEntry {
                                    item_id: 1014,
                                    association_count: 2,
                                    associations: [
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: true,
                                            property_index: 6,
                                        },
                                        ItemPropertyAssociationBoxEntryAssociation {
                                            essential: false,
                                            property_index: 7,
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
            ),
            grpl: None,
            unknown_boxes: [],
        },
    ),
    styp: [],
    sidx: [],
    ssix: [],
    prft: [],
    unknown_boxes: [],
    mfra: None,
}
